src/gen_server2.erl

```diff
@@ -576,10 +576,11 @@ init_it(Starter, Parent, Name0, Mod, Args, Options) ->
         {ok, State, Timeout, Backoff = {backoff, _, _, _}, Mod1} ->
             Backoff1 = extend_backoff(Backoff),
             proc_lib:init_ack(Starter, {ok, self()}),
-            loop(GS2State #gs2_state { mod           = Mod1,
-                                       state         = State,
-                                       time          = Timeout,
-                                       timeout_state = Backoff1 });
+            loop(find_prioritisers(
+                  GS2State #gs2_state { mod           = Mod1,
+                                        state         = State,
+                                        time          = Timeout,
+                                        timeout_state = Backoff1 }));
```

Comment:
The problem is that the `Mod` we pass to the first call to `find_prioritisers/1` is not the same `Mod` passed in the second call. The bug was precisely there, not calling `find_prioritisers/1` for the new `Mod` that was swapped by the `gen_server2` callback.

`rabbit_prequeue` starts as the `gen_server2` callback being `Mod`, then `Mod:init/1` is called, which returns a new module, in our case it could be `rabbit_amqqueue_process`. The module `rabbit_amqqueue_process` will become the new `gen_server2`, so we have to call `find_prioritisers/1` again